src/goto-instrument/contracts/dynamic-frames/dfcc_swap_and_wrap.cpp

```diff
@@ -0,0 +1,313 @@
+              << messaget::eom;
+
+  // all code generation needs to run on functions with unmodified signatures
+  const irep_idt &wrapper_id = function_id;
+  const irep_idt wrapped_id =
+    id2string(wrapper_id) + "_wrapped_for_contract_checking";
+  utils.wrap_function(wrapper_id, wrapped_id);
+
+  // wrapper body
+  goto_programt body;
+
+  const auto &wrapper_symbol = utils.get_function_symbol(wrapper_id);
+
+  auto check_started = utils
+                         .create_static_symbol(
+                           bool_typet(),
+                           id2string(function_id),
+                           "__contract_check_in_progress",
+                           wrapper_symbol.location,
+                           wrapper_symbol.mode,
+                           wrapper_symbol.module,
+                           false_exprt())
+                         .symbol_expr();
+
+  auto check_completed = utils
+                           .create_static_symbol(
+                             bool_typet(),
+                             id2string(function_id),
+                             "__contract_checked_once",
+                             wrapper_symbol.location,
+                             wrapper_symbol.mode,
+                             wrapper_symbol.module,
+                             false_exprt())
+                           .symbol_expr();
+
+  auto check_started_goto =
+    body.add(goto_programt::make_incomplete_goto(check_started));
+
+  // At most a single top level call to the checked function in any execution
+
+  // Recursive calls within a contract check correspond to
+  // `check_started && !check_completed` and are allowed.
+
+  // Any other call occuring with `check_completed` true is forbidden.
+  source_locationt sl;
+  sl.set_comment(
+    "Only a single top-level call to function " + id2string(function_id) +
+    " when checking contract " + id2string(contract_id));
+  body.add(goto_programt::make_assertion(not_exprt(check_completed), sl));
+  body.add(goto_programt::make_assignment(check_started, true_exprt()));
+
+  const auto write_set_symbol = utils.create_new_parameter_symbol(
+    function_id,
+    "__write_set_to_check",
+    library.dfcc_type[dfcc_typet::CAR_SET_PTR]);
+
+  contract_handler.add_contract_instructions(
+    dfcc_contract_modet::CHECK,
+    wrapper_id,
+    wrapped_id,
+    contract_id,
+    write_set_symbol,
+    body,
+    function_pointer_contracts);
+
+  body.add(goto_programt::make_assignment(check_completed, true_exprt()));
+  body.add(goto_programt::make_assignment(check_started, false_exprt()));
+
+  // unconditionally Jump to the end after the check
+  auto goto_end_function = body.add(goto_programt::make_incomplete_goto());
+
+  // Jump to the replacement section if check already in progress
+  auto contract_replacement_label = body.add(goto_programt::make_skip());
+  check_started_goto->complete_goto(contract_replacement_label);
```

Comment:
Is this an unconditional `goto` to the next instruction?!

Reply:
no it is not,  the control flow is is a if-then-else type of structure

 ```c
 IF __contract_check_in_progress GOTO else_branch;
 ASSERT !__contract_checked_once "only a single top-level called allowed";
 __contract_check_in_progress = true;
    ...... some instructions ......
 __contract_checked_once = true;
 __contract_check_in_progress = false;
 GOTO end;
 else_branch:
 // if allow_recursive_calls
    ...... some instructions ......
 // if !allow_recursive_calls
 ASSERT false, "no recursive calls";
 ASSUME false;
 end:
 END_FUNCTION;
 ```